src/coreclr/src/jit/hwintrinsic.cpp

```diff
@@ -643,37 +639,85 @@ GenTree* Compiler::impHWIntrinsic(NamedIntrinsic        intrinsic,
         assert(sizeBytes != 0);
     }
 
-    // NOTE: The following code assumes that for all intrinsics
-    // taking an immediate operand, that operand will be last.
-    if (sig->numArgs > 0 && HWIntrinsicInfo::isImmOp(intrinsic, impStackTop().val))
+    GenTree* immOp = nullptr;
+
+#ifdef TARGET_ARM64
+    if (intrinsic == NI_AdvSimd_Insert)
+    {
+        assert(sig->numArgs == 3);
+        immOp = impStackTop(1).val;
+        assert(HWIntrinsicInfo::isImmOp(intrinsic, immOp));
+    }
+    else
+#endif
+        if ((sig->numArgs > 0) && HWIntrinsicInfo::isImmOp(intrinsic, impStackTop().val))
+    {
+        // NOTE: The following code assumes that for all intrinsics
+        // taking an immediate operand, that operand will be last.
+        immOp = impStackTop().val;
+    }
+
+    int immUpperBound = 0;
+
+    if (immOp != nullptr)
     {
-        GenTree* lastOp = impStackTop().val;
-        // The imm-HWintrinsics that do not accept all imm8 values may throw
-        // ArgumentOutOfRangeException when the imm argument is not in the valid range
+        bool constImm = false;
+
         if (!HWIntrinsicInfo::HasFullRangeImm(intrinsic))
         {
-            if (!mustExpand && lastOp->IsCnsIntOrI() &&
-                !HWIntrinsicInfo::isInImmRange(intrinsic, (int)lastOp->AsIntCon()->IconValue()))
+            bool constImmOutOfRange = false;
+
+#if defined(TARGET_ARM64)
+            const var_types elemType = getBaseTypeFromArgIfNeeded(intrinsic, clsHnd, sig, baseType);
+            const int       simdSize = HWIntrinsicInfo::lookupSimdSize(this, intrinsic, sig);
+            immUpperBound            = HWIntrinsicInfo::lookupImmUpperBound(intrinsic, simdSize, elemType);
+
+            if (immOp->IsCnsIntOrI())
             {
-                return nullptr;
+                const int ival = (int)immOp->AsIntCon()->IconValue();
+
+                constImm           = true;
+                constImmOutOfRange = ((ival < 0) || (ival >= immUpperBound));
             }
-        }
 
-        if (!lastOp->IsCnsIntOrI())
-        {
-            if (HWIntrinsicInfo::NoJmpTableImm(intrinsic))
+#elif defined(TARGET_XARCH)
+            if (immOp->IsCnsIntOrI())
             {
-                return impNonConstFallback(intrinsic, retType, baseType);
+                const int ival = (int)immOp->AsIntCon()->IconValue();
+
+                constImm           = true;
+                constImmOutOfRange = !HWIntrinsicInfo::isInImmRange(intrinsic, ival);
+            }
+            else
+            {
+                immUpperBound = HWIntrinsicInfo::lookupImmUpperBound(intrinsic);
```

Comment:
We generally need the other bits of info anyways. Any reason we don't just pass the unnecessary args on x86 (or define an inline function so they can get optimized away) so this code doesn't need as many ifdefs?

Reply:
Perhaps - the original intent was to come back to this part of the code later when I start implementing `*BySelectedScalar` or even arithmetic intrinsics accepting immediate - I need to see more different cases before this can be refactored to something more readable. By if you insist, I can work on this now.

Reply:
I was just thinking we could trivially change the x86 signature as well and continue sharing one code path, rather than needing to insert an `#ifdef`

I don't think it needs to be anything significantly more than that right now...

Reply:
I will try to simpilfy - agree that this is the most confusing part of the changes.

Reply:
@tannergooding Can you take a look at `impHWIntrinsic` after I pushed two last commits?